v2/yarpcgrpc/integration_test.go

```diff
@@ -21,521 +21,391 @@
-		assert.NoError(t, err)
-		e.KeyValueYARPCServer.SetNextError(status.Error(codes.FailedPrecondition, "bar 1"))
-		value, err := e.GetValueGRPC(context.Background(), "foo")
-		assert.Equal(t, "bar", value)
-		assert.Equal(t, status.Error(codes.FailedPrecondition, "bar 1"), err)
+				<-start
+				err := client.Call(ctx, "test-procedure", &testEchoRequest{Message: msg}, &res)
+
+				lock.Lock()
+				errs = multierr.Combine(errs, err)
+				lock.Unlock()
+
+				wg.Done()
+			}(i)
+		}
+		close(start)
+		wg.Wait()
+		require.NoError(t, errs)
 	})
 }
 
-func TestYARPCMaxMsgSize(t *testing.T) {
-	t.Parallel()
-	value := strings.Repeat("a", defaultServerMaxRecvMsgSize*2)
-	te := testEnvOptions{}
-	te.do(t, func(t *testing.T, e *testEnv) {
-		assert.Equal(t, yarpcerrors.CodeResourceExhausted, yarpcerrors.FromError(e.SetValueYARPC(context.Background(), "foo", value)).Code())
-	})
-	te = testEnvOptions{
-		TransportOptions: []TransportOption{
-			ClientMaxRecvMsgSize(math.MaxInt32),
-			ClientMaxSendMsgSize(math.MaxInt32),
-			ServerMaxRecvMsgSize(math.MaxInt32),
-			ServerMaxSendMsgSize(math.MaxInt32),
-		},
-	}
-	te.do(t, func(t *testing.T, e *testEnv) {
-		assert.NoError(t, e.SetValueYARPC(context.Background(), "foo", value))
-		getValue, err := e.GetValueYARPC(context.Background(), "foo")
-		assert.NoError(t, err)
-		assert.Equal(t, value, getValue)
-	})
+type testEchoRequest struct {
+	Message string
+	Error   string
 }
 
-func TestLargeEcho(t *testing.T) {
-	t.Parallel()
-	value := strings.Repeat("a", 32768)
-	te := testEnvOptions{}
-	te.do(t, func(t *testing.T, e *testEnv) {
-		assert.NoError(t, e.SetValueYARPC(context.Background(), "foo", value))
-		getValue, err := e.GetValueYARPC(context.Background(), "foo")
-		assert.NoError(t, err)
-		assert.Equal(t, value, getValue)
-	})
+type testEchoResponse struct {
+	Message string
 }
 
-func TestApplicationErrorPropagation(t *testing.T) {
-	t.Parallel()
-	te := testEnvOptions{}
-	te.do(t, func(t *testing.T, e *testEnv) {
-		response, err := e.Call(
-			context.Background(),
-			"GetValue",
-			&examplepb.GetValueRequest{Key: "foo"},
-			protobuf.Encoding,
-			transport.Headers{},
-		)
-		require.Equal(t, yarpcerrors.NotFoundErrorf("foo"), err)
-		require.True(t, response.ApplicationError)
-
-		response, err = e.Call(
-			context.Background(),
-			"SetValue",
-			&examplepb.SetValueRequest{Key: "foo", Value: "hello"},
-			protobuf.Encoding,
-			transport.Headers{},
-		)
-		require.NoError(t, err)
-		require.False(t, response.ApplicationError)
-
-		response, err = e.Call(
-			context.Background(),
-			"GetValue",
-			&examplepb.GetValueRequest{Key: "foo"},
-			"bad_encoding",
-			transport.Headers{},
-		)
-		require.True(t, yarpcerrors.IsInvalidArgument(err))
-		require.False(t, response.ApplicationError)
-	})
+func testEchoHandler(_ context.Context, request *testEchoRequest) (*testEchoResponse, error) {
+	var err error
+	if request.Error != "" {
+		err = errors.New(request.Error)
```

Comment:
Should we return `nil, errors.New(request.Error)` in this case so that it better reflects the idiom of non-nil errors resulting in `nil` responses?

Reply:
yup